serial/tools/list_ports_linux.py

```diff
@@ -61,54 +62,60 @@ def __init__(self, device):
             self.hwid = self.read_line(self.device_path, 'id')
         elif self.subsystem == 'amba':  # raspi
             self.description = self.name
             self.hwid = os.path.basename(self.device_path)
 
         if is_link:
-            self.hwid += ' LINK={}'.format(device)
+            self.hwid += f' LINK={device}'
 
-    def read_line(self, *args):
+    @staticmethod
+    def read_line(arg: str, *args: str) -> str | None:
         """\
         Helper function to read a single line from a file.
         One or more parameters are allowed, they are joined with os.path.join.
-        Returns None on errors..
+        Returns None on errors.
         """
         try:
-            with open(os.path.join(*args)) as f:
+            with open(os.path.join(arg, *args)) as f:
                 line = f.readline().strip()
             return line
-        except IOError:
+        except OSError:
             return None
 
 
-def comports(include_links=False):
+_known_device_patterns = (
+    # These patterns are tested in the test suite.
+    '/dev/ttyS*',     # built-in serial ports
+    '/dev/ttyUSB*',   # usb-serial with own driver
+    '/dev/ttyACM*',   # usb-serial with CDC-ACM profile
+    #
+    # These patterns are currently untested. If you have one of these devices,
+    # run the pyserial-generate-test-asset command and file an issue.
+    '/dev/ttyXRUSB*', # xr-usb-serial port exar (DELL Edge 3001)
+    '/dev/ttyAMA*',   # ARM internal port (raspi)
+    '/dev/rfcomm*',   # BT serial devices
+    '/dev/ttyAP*',    # Advantech multi-port serial controllers
+    '/dev/ttyGS*',    # https://www.kernel.org/doc/Documentation/usb/gadget_serial.txt
+)
+
+
+def comports(include_links: bool = False) -> list[SysFS]:
     devices = set()
-    devices.update(glob.glob('/dev/ttyS*'))     # built-in serial ports
-    devices.update(glob.glob('/dev/ttyUSB*'))   # usb-serial with own driver
-    devices.update(glob.glob('/dev/ttyXRUSB*')) # xr-usb-serial port exar (DELL Edge 3001)
-    devices.update(glob.glob('/dev/ttyACM*'))   # usb-serial with CDC-ACM profile
-    devices.update(glob.glob('/dev/ttyAMA*'))   # ARM internal port (raspi)
-    devices.update(glob.glob('/dev/rfcomm*'))   # BT serial devices
-    devices.update(glob.glob('/dev/ttyAP*'))    # Advantech multi-port serial controllers
-    devices.update(glob.glob('/dev/ttyGS*'))    # https://www.kernel.org/doc/Documentation/usb/gadget_serial.txt
+    for pattern in _known_device_patterns:
+        devices.update(glob.glob(pattern))
 
     if include_links:
         devices.update(list_ports_common.list_links(devices))
-    return [info
-            for info in [SysFS(d) for d in devices]
-            if info.subsystem != "platform"]    # hide non-present internal serial ports
-
-# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
-# test
-if __name__ == '__main__':
-    for info in sorted(comports()):
-        print("{0}: {0.subsystem}".format(info))
```

Comment:
Was it neccesary to remove this?  As much as formal testing is done with pytest I often like little test harnesses like this myself.

Reply:
It's not necessary for this PR. However, these little chunks of untestable code need to be removed.

The existing `pyserial-ports` command is where this should live, which does all of the same work using tested code. It just needs to be extended to include the subsystem in its output.